indispensable-cosef/src/commonMain/kotlin/at/asitplus/signum/indispensable/cosef/CoseSignedSerializer.kt

```diff
@@ -1,152 +1,97 @@
-                2 -> payload = decodeNullableSerializableElement(jsonDescriptor, 2, ByteArraySerializer())
-                3 -> signature = decodeSerializableElement(jsonDescriptor, 3, ByteArraySerializer())
-                else -> break
-            }
-        }
-        toTypedObject(payload, protectedHeader, unprotectedHeader, signature)
-    }
-
-    private fun toTypedObject(
-        payload: ByteArray?,
-        protectedHeader: ByteStringWrapper<CoseHeader>,
+    private fun ByteArray.toSignature(
+        protectedHeader: CoseHeader,
         unprotectedHeader: CoseHeader?,
-        signature: ByteArray,
-    ): CoseSigned<P> = runCatching {
-        val typedPayload: P? = payload?.let {
-            coseCompliantSerializer.decodeFromByteArray(parameterSerializer, it)
-        }
-        CoseSigned(protectedHeader, unprotectedHeader, typedPayload, signature)
-    }.getOrElse {
-        @Suppress("UNCHECKED_CAST")
-        (CoseSigned(protectedHeader, unprotectedHeader, payload as P, signature))
+    ): CryptoSignature.RawByteEncodable =
+        if (protectedHeader.usesEC() ?: unprotectedHeader?.usesEC() ?: (size < 2048))
+            CryptoSignature.EC.fromRawBytes(this)
+        else CryptoSignature.RSAorHMAC(this)
+
+    private fun ByteArray?.toNullablePayload(): P? = when (this) {
+        null -> null
+        else -> if (this.isEmpty()) null else toTypedPayload()
     }
 
-    override fun serialize(encoder: Encoder, value: CoseSigned<P>) {
-        (if (encoder is CborEncoder) descriptor else jsonDescriptor).let { descriptor ->
-            encoder.encodeStructure(descriptor) {
-                encodeSerializableElement(
-                    descriptor,
-                    0,
-                    ByteStringWrapperSerializer(CoseHeader.serializer()),
-                    value.protectedHeader
-                )
-                encodeNullableSerializableElement(descriptor, 1, CoseHeader.serializer(), value.unprotectedHeader)
-
-                when(value.payload) {
-                    is ByteArray -> encodeNullableSerializableElement(
-                        descriptor,
-                        2,
-                        ByteArraySerializer(),
-                        value.payload as ByteArray
-                    )
-                    is Nothing? -> encodeNullableSerializableElement(descriptor, 2, parameterSerializer, value.payload)
-
-                    else -> encodeNullableSerializableElement(
-                        buildTag24SerialDescriptor(),
-                        2,
-                        ByteStringWrapperSerializer(parameterSerializer),
-                        ByteStringWrapper(value.payload!!)
-                    )
-                }
-                encodeSerializableElement(descriptor, 3, ByteArraySerializer(), value.rawSignature)
-            }
+    private fun ByteArray.toTypedPayload(): P =
+        if (parameterSerializer == ByteArraySerializer()) {
+            typed()
```

Comment:
please inline this here, since here is the only legal use of such a cast

Reply:
Fixed that